packages/compass-preferences-model/src/preferences.ts

```diff
@@ -157,6 +158,29 @@ function deriveNetworkTrafficOptionState<K extends keyof AllPreferences>(
   });
 }
 
+/** Helper for defining how to derive value/state for feature-restricting preferences */
+function deriveFeatureRestrictingOptionsState<K extends keyof AllPreferences>(
+  property: K
+): DeriveValueFunction<boolean> {
+  // TODO: COMPASS-6063
+  // Check maxTimeMS as restricting option,
+  return (v, s) => ({
+    value:
+      v(property) &&
+      v('enableShell') &&
+      !v('protectConnectionStrings') &&
+      !v('readOnly'),
+    state:
+      s(property) ??
+      s('protectConnectionStrings') ??
+      (v('protectConnectionStrings') ? 'derived' : undefined) ??
+      s('readOnly') ??
+      (v('readOnly') ? 'derived' : undefined) ??
+      s('enableShell') ??
+      (v('enableShell') ? undefined : 'derived'),
```

Comment:
Thinking about this, I probably slightly messed up the other definitions here – this should maybe be more like


```suggestion
      s(property) ??
      (v('protectConnectionStrings') ? (s('protectConnectionStrings') ?? 'derived') : undefined) ??
      (v('readOnly') ? (s('readOnly') ?? 'derived') : undefined) ??
      (v('enableShell') ? undefined: (s('enableShell') ?? 'derived')),
```

(and similarly for `deriveNetworkTrafficOptionState` / `deriveReadOnlyOptionState` …)

Because somebody can of course force the default value of an option through the command line or global config file, i.e. `--no-protect-connection-strings`, which doesn’t change Compass’s behavior compared to the default and therefore probably also shouldn’t affect `enableDevTools`, but does set the state of `protectConnectionStrings` to `set-cli`.